multierror.go

```diff
@@ -49,3 +50,67 @@ func (e *Error) GoString() string {
 func (e *Error) WrappedErrors() []error {
 	return e.Errors
 }
+
+// Unwrap returns an error from Error (or nil if there are no errors).
+// This error returned will further support Unwrap to get the next error,
+// etc. The order will match the order of Errors in the multierror.Error
+// at the time of calling.
+//
+// The resulting error supports errors.As/Is/Unwrap so you can continue
+// to use the stdlib errors package to introspect further.
+//
+// This will perform a shallow copy of the errors slice. Any errors appended
+// to this error after calling Unwrap will not be available until a new
+// Unwrap is called on the multierror.Error.
+func (e *Error) Unwrap() error {
+	// If we have no errors then we do nothing
+	if e == nil || len(e.Errors) == 0 {
+		return nil
+	}
+
+	// If we have exactly one error, we can just return that directly.
+	if len(e.Errors) == 1 {
+		return e.Errors[0]
+	}
+
+	// Shallow copy the slice
+	errs := make([]error, len(e.Errors))
+	copy(errs, e.Errors)
+	return chain(errs)
+}
+
+// chain implements the interfaces necessary for errors.Is/As/Unwrap to
+// work in a deterministic way with multierror. A chain tracks a list of
+// errors while accounting for the current represented error. This lets
+// Is/As be meaningful.
+//
+// Unwrap returns the next error. In the cleanest form, Unwrap would return
+// the wrapped error here but we can't do that if we want to properly
+// get access to all the errors. Instead, users are recommended to use
+// Is/As to get the correct error type out.
+type chain []error
+
+// Error implements the error interface
+func (e chain) Error() string {
```

Comment:
Though it can't happen I'd argue for a len check on this and As/Is since this is used in some pretty low level applications. Some future change could make it possible and this is a lib I don't mind being too defensive.

Reply:
I wanted to avoid the len check since this is unexported and there are guards all around. I think proper behavior is a panic in that case cause it represents some significant bugs. I tested around all the current edges I can think of that this could happen.

Reply:
And maybe a test explicitly calling:
```
chain([]error{}).Error()
```

Reply:
I'm fine with that then!

Reply:
This should be impossible, I think instead documenting a precondition that chain must be on a non-empty error slice is the way to go.